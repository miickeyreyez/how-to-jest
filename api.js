import fetch from 'isomorphic-fetch';

// eslint-disable-next-line max-len
const api = async () => fetch('https://jsonplaceholder.typicode.com/todos/1')
  .then(response => response.json())
  .then(json => json);

export default api;
