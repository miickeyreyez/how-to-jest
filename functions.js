const sqrt = (a, b) => a ** b;

const anotherHelloWorld = () => 'Hello World!';

export {
  sqrt,
  anotherHelloWorld,
};
