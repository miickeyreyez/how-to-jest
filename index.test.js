import {
  sum, isOddNumber,
  getUserFromApi, helloWorld, customError, createFile,
} from '.';
import api from './api';
import * as functions from './functions';
import MyCustomError from './error';

jest.mock('./api');
jest.mock('fs');

// eslint-disable-next-line max-len
functions.anotherHelloWorld = jest.fn().mockImplementation(() => 'Mock hello world');

const userStub = {
  completed: true,
  id: 1001,
  title: 'My own title',
  userId: 1001,
};

api.mockImplementation(async flag => new Promise((resolve, reject) => {
  if (flag) resolve(userStub);
  reject(new MyCustomError());
}));

test('helloWorld function returns Hello World string', () => {
  expect(helloWorld()).toEqual('Hello World!');
});

test('anotherHelloWorld function returns my mock Hello World string', () => {
  expect(functions.anotherHelloWorld()).toEqual('Mock hello world');
});

test('helloWorld function returns Hello World string', () => {
  expect(helloWorld()).not.toBe('Hello World');
});

describe('sum function', () => {
  it('should return the sum of two numbers', () => {
    expect(sum(2, 2)).toBe(4);
  });
  it('should return the sum of two numbers', () => {
    expect(sum(2, 2)).toEqual(4);
  });
});

describe('isOdddNumber function', () => {
  it('should return true if number is odd', () => {
    expect(isOddNumber(4)).toBe(true);
  });

  it('should return true if number is odd', () => {
    expect(isOddNumber(5)).toBe(false);
  });
});

describe('getUserFromApi function', () => {
  it('should return a user from api (Method 1)', async () => {
    const a = await getUserFromApi();
    expect(a).toMatchSnapshot();
  });

  it('should return a user from api (Method 2)', async () => {
    expect.assertions(1);
    await expect(getUserFromApi()).resolves.toMatchSnapshot();
  });
});

describe('api function', () => {
  it('should return a user from api case 1 (Mock)', async () => {
    expect.assertions(1);
    await expect(api(true)).resolves.toMatchSnapshot();
  });

  it('should return a user from api case 2 (Mock)', async () => {
    expect.assertions(1);
    await expect(api(true)).resolves.toMatchObject(userStub);
  });

  it('should return a user from api case 3 (Mock)', async () => {
    expect.assertions(1);
    await expect(api(true)).resolves.toMatchObject({
      id: expect.any(Number),
    });
  });

  it('should thrown an exception', async () => {
    expect.assertions(1);

    return api(false).catch((error) => {
      expect(error).toEqual(new MyCustomError());
    });
  });
});

test('Custom exception', () => {
  expect(() => customError()).toThrowError(new MyCustomError());
});

describe('createFile function', () => {
  it('should generate file successfully', () => {
    expect(() => createFile('filePath', 'fileContent')).not.toThrow();
  });

  it('should throw an exception when the file path is not provided', () => {
    expect(() => createFile('filePath', null))
      .toThrowError(new MyCustomError());
  });

  it('should throw an exception when the file content is not null', () => {
    expect(() => createFile(null, 'fileContent'))
      .toThrowError(new MyCustomError());
  });
});
