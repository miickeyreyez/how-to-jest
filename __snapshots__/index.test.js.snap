// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`api function should return a user from api case 1 (Mock) 1`] = `
Object {
  "completed": true,
  "id": 1001,
  "title": "My own title",
  "userId": 1001,
}
`;

exports[`getUserFromApi function should return a user from api (Method 1) 1`] = `
Object {
  "completed": false,
  "id": 1,
  "title": "delectus aut autem",
  "userId": 1,
}
`;

exports[`getUserFromApi function should return a user from api (Method 2) 1`] = `
Object {
  "completed": false,
  "id": 1,
  "title": "delectus aut autem",
  "userId": 1,
}
`;
